File: 734. Sentence Similarity.py
class Solution:
    def areSentencesSimilar(self, words1: List[str], words2: List[str], pairs: List[List[str]]) -> bool:
        if len(words1) != len(words2):
            return False
        pairset = set([tuple(i) for i in pairs])
        
        return all(w1 == w2 or (w1, w2) in pairset or (w2, w1) in pairset for w1, w2 in zip(words1, words2))

class Solution:
    def areSentencesSimilar(self, words1: List[str], words2: List[str], pairs: List[List[str]]) -> bool:
        if len(words1) != len(words2):
            return False
        dic = collections.defaultdict(set)
        for k, v in pairs:
            dic[k].add(v)
        
        for w1, w2 in zip(words1, words2):
            if w1 == w2:
                continue
            if w1 not in dic[w2] and w2 not in dic[w1]:
                return False
        return True

File: 734. Sentence Similarity.java
class Solution {
    public boolean areSentencesSimilar(String[] words1, String[] words2, List<List<String>> pairs) {
        if (words1.length != words2.length) {
            return false;
        }
        HashMap<String, Set<String>> map = new HashMap<>();
        for (List<String> l : pairs) {
            String w1 = l.get(0);
            String w2 = l.get(1);
            if (map.containsKey(w1)) {
                map.get(w1).add(w2);
            } else {
                Set<String> set = new HashSet<>();
                set.add(w2);
                map.put(w1, set);
            }
            if (map.containsKey(w2)) {
                map.get(w2).add(w1);
            } else {
                Set<String> set = new HashSet<>();
                set.add(w1);
                map.put(w2, set);
            }
        }
        for (int i=0; i<words1.length; i++) {
            if (words1[i].equals(words2[i])) continue;
            if (!map.getOrDefault(words1[i], new HashSet<>()).contains(words2[i]) ||
               !map.getOrDefault(words2[i], new HashSet<>()).contains(words1[i])) {
                return false;
            }
        }
        return true;
    }
}